### PR TITLE
feat(security): handle grant constraints, memory policy, redaction boundary tests

### LINKED ISSUE #74: Add redaction boundary tests for frames, raw results, and traces
## Context
`agent-kernel` models raw driver results, safe `Frame` objects, handles, provenance, and action traces. The intended boundary is clear: raw driver output should not reach the LLM-safe frame unless explicitly allowed for safe admin paths.

## Problem
As the execution, audit, and MCP layers grow, it becomes easier for raw data or sensitive arguments to leak through frames, traces, dry-run results, or error messages. The repo needs focused tests that lock down the redaction/firewall boundary.

## Scope
- Add regression tests for raw result to frame conversion/redaction.
- Cover allowed fields, PII-like fields, secret-like values, handle-only output, raw response mode, and trace/audit metadata.
- Verify warnings/provenance are preserved without exposing raw sensitive values.
- Add small docs/comments describing the expected boundary.

## Out of scope
- Do not implement a full secret scanner.
- Do not change authorization semantics except where a test reveals a bug.
- Do not add real credentials or personal data to fixtures.
- Do not remove admin/raw modes if they are intentional; document and test their constraints.

## Suggested implementation
Inspect firewall/redaction code, `Frame`, `RawResult`, `Handle`, `ActionTrace`, and tests around response modes. Use synthetic records with fields such as `email`, `token`, `internal_notes`, and `public_status`. Assert that only allowed data reaches the frame and that traces contain safe metadata.

## Acceptance criteria
- [ ] Tests prove raw sensitive fields are excluded from default LLM-safe frames.
- [ ] Tests prove handle-only or redacted outputs preserve useful references without raw content.
- [ ] Tests cover trace/audit metadata so sensitive args are not leaked there by default.
- [ ] Raw/admin response behavior is explicitly tested or documented.
- [ ] Existing tests pass.

## Tests
Add unit tests for frame construction/firewall behavior and trace generation. Include negative assertions for synthetic secret values and PII-like values.

## Documentation
Update architecture or firewall docs with a brief statement of what may appear in frames/traces versus what must stay in handles/raw storage.

## Dependencies
Related to #73 if traces are extended. Can be implemented independently for current models.

## AI agent notes
Start by locating firewall/redaction tests and response mode handling. Use obviously fake secrets. Prefer exact assertions over broad snapshots. Be careful that test failure output does not print fake secret strings unnecessarily.

### LINKED ISSUE #75: Add policy checks for memory read and write actions
## Context
`agent-kernel` models capabilities, policy decisions, frames, handles, and audit traces for tool-using agents. As agents gain persistent memory, memory read/write operations become security-relevant actions that should be represented and governed like other capabilities.

## Problem
Reading memory can expose sensitive past context. Writing memory can create durable false assumptions, leak secrets into future sessions, or pollute project-level context. Without explicit memory capabilities, policy and audit layers cannot distinguish safe ephemeral context use from durable memory access.

## Scope
- Model memory read and memory write as capability/action types or equivalent policy inputs.
- Allow policy checks to distinguish read vs write, memory scope, sensitivity, and durability.
- Add audit trace entries for memory access decisions.
- Add examples/tests for allowed project memory read, denied sensitive memory read, allowed safe memory write, and ask/deny for durable memory write if supported.
- Keep the implementation generic and storage-agnostic.

## Out of scope
- Do not implement a memory database or vector store.
- Do not integrate with a specific memory product.
- Do not add background memory capture.
- Do not remove existing generic capability behavior.

## Suggested implementation
Inspect capability request/decision models, policy engine, action trace, and the new intent/scope issue #72. Consider using existing capability names such as `memory.read` and `memory.write` or adding a small helper model for memory actions. Preserve backward compatibility.

## Acceptance criteria
- [ ] Memory read/write actions can be represented in policy evaluation.
- [ ] Policy can distinguish memory scope and sensitivity.
- [ ] Audit traces record memory access decisions without exposing raw sensitive memory by default.
- [ ] Tests cover allow and deny paths for read and write.
- [ ] Docs include a storage-agnostic memory policy example.

## Tests
Add unit tests for policy evaluation of memory read/write actions and audit trace output. Include a redaction test for sensitive memory payloads.

## Documentation
Add a short example policy section for memory actions, including project-scoped memory and sensitive/personal memory.

## Dependencies
Pairs naturally with #72, #73, and #74. Should align with `weaver-spec` MemoryArtifact/SessionHandoff contract issue #56.

## AI agent notes
Start from existing capability and policy models. Keep this storage-neutral. Avoid logging raw memory text in traces. Treat memory writes as potentially higher risk than memory reads because they persist into future sessions.

### LINKED ISSUE #76: security: enforce grant constraints during handle expansion
## Context
After inspecting the repository and existing issues, `agent-kernel` already has policy, trace, handle, and OTel-oriented work. I did not find an existing issue specifically ensuring that grant constraints are persisted with handles and rechecked when handles are expanded.

## Problem
A handle should not expand beyond the constraints attached to the original grant. Limits such as allowed fields, max rows/items, expiry, or scope need to remain enforceable after the initial policy decision.

## Scope
- Persist relevant grant constraints with handles when they are created.
- Recheck those constraints during handle expansion.
- Cover max rows/items, allowed fields, expiry, and scope constraints where those concepts exist in the current model.
- Return predictable denial/error results when expansion violates constraints.
- Add tests for allowed and denied expansion paths.

## Out of scope
- Do not redesign the full policy model.
- Do not add a new storage backend.
- Do not require OpenTelemetry or external logging.
- Do not change unrelated capability execution behavior.

## Suggested implementation
Inspect handle creation/expansion code, grant/capability models, policy decision code, firewall/redaction code, and trace/audit behavior. Likely modules to inspect first include handle store/manager, policy/grant models, expansion APIs, and tests around handles.

## Acceptance criteria
- [ ] Expansion respects max rows/items or equivalent limits when available.
- [ ] Expansion respects allowed fields or equivalent projection constraints when available.
- [ ] Expired or out-of-scope handles are denied predictably.
- [ ] Denied expansion returns a stable error/denial shape.
- [ ] Tests cover allowed and denied expansion.

## Tests
Add unit tests for a handle that expands successfully, a handle exceeding row/item limits, a handle requesting disallowed fields, and an expired/out-of-scope handle if the model supports it.

## Documentation
Document the handle expansion security invariant and the expected denial behavior. A short note in security docs or handle API docs is enough.

## Dependencies
None.

## AI agent notes
Preserve backwards compatibility where possible. If existing handles do not carry constraint metadata, add a migration-safe default or versioned shape. Keep denial results machine-readable so future policy tests can assert behavior without matching fragile strings.

### PR DESCRIPTION
**Pull request by dgenio**
﻿## Summary

Implements the security-boundary group identified during triage: **#76**
(handle grant-constraint enforcement), **#75** (memory action policy), and
**#74** (redaction boundary tests). One PR because all three changes touch
the same seam ÔÇö what data is allowed to flow from driver/store outward, and
how policy + trace + firewall enforce it.

This PR also closed two related issues during triage: **#73** (decision
trace format) and **#77** (stable denial reason codes) ÔÇö both implemented
by PR #78 but the auto-close didn't fire. Each has a closing comment
explaining the evidence.

## What changed

**Handle grant-constraint enforcement (#76)**
- `Handle` now carries `principal_id` and `constraints` from the original
  grant, persisted at creation time by `HandleStore.store`.
- `HandleStore.expand` rechecks against the requested expand query:
  `max_rows`, `allowed_fields`, `scope` filter, and principal binding.
  Violations raise `HandleConstraintViolation` (new exception, exported
  from `agent_kernel`) with stable `reason_code`.
- An unscoped `expand` applies `allowed_fields` as the default projection
  and AND-merges `scope` into the filter, so a caller can't bypass either
  constraint by omitting it.
- `Kernel.expand` gained an optional `principal: Principal` argument that
  is forwarded for the principal-binding check.

**Memory action policy (#75)**
- New `SensitivityTag.MEMORY`.
- `DefaultPolicyEngine.evaluate` adds a MEMORY branch *after* existing
  sensitivity checks (per the "rule placement matters" trap in
  `docs/agent-context/invariants.md`):
  - `memory.read` with `scope.memory_scope == "sensitive"` requires
    `memory_reader_sensitive` or `admin`.
  - `memory.write` / DESTRUCTIVE memory always requires `memory_writer`
    or `admin`.
  - Project-scoped reads are allowed by default.
- `explain()` lists the same conditions with stable `reason_code`s.
- Kernel-side redaction in `ActionTrace.args`: for capabilities whose ID
  starts with `memory.`, payload-like keys (`payload`, `content`, `value`,
  `memory`, `text`, `body`) are replaced with `"[REDACTED]"`. Keys are
  preserved so audit can still confirm the action.

**Redaction boundary tests (#74)**
- New `tests/test_firewall_boundary.py` with 8 negative-assertion tests:
  - Synthetic secret/PII patterns never appear in summary/table frames.
  - `allowed_fields` strips disallowed columns AND their key names.
  - `handle_only` frame carries only a Handle reference (no rows, no raw).
  - Non-admin asking for `raw` is silently downgraded to summary;
    raw_data stays `None`.
  - Admin in raw mode receives raw data but inline secret patterns
    (Bearer tokens, API keys) are still scrubbed by `redact()`.
  - `ActionTrace.args` for memory capabilities redacts payload-like keys.
  - The redaction is scoped to `memory.*` ÔÇö non-memory capabilities are
    untouched.

**New stable denial codes** (added to `agent_kernel.policy_reasons`):
- `HANDLE_CONSTRAINT_VIOLATION` (`handle_constraint_violation`)
- `HANDLE_PRINCIPAL_MISMATCH` (`handle_principal_mismatch`)
- `MEMORY_WRITE_REQUIRES_WRITER` (`memory_write_requires_writer`)
- `MEMORY_SENSITIVE_READ_DENIED` (`memory_sensitive_read_denied`)

**Docs**
- `docs/security.md`: new "Handle expansion boundary" and "Memory actions"
  sections; expanded threat-model table; expanded confused-deputy section
  to cover handles.
- `docs/architecture.md`: MEMORY entry in `DefaultPolicyEngine` rule list;
  four new rows in the reason-code table.
- `CHANGELOG.md`: `[Unreleased]` documents the three changes.

## Why

From triage: handle expansion was the load-bearing seam where over-broad
data could escape the original grant; memory writes persist into future
sessions so they need an explicit policy layer; and the redaction
boundary lacked focused negative-assertion tests pinning the contract.
The three issues share the same modules (`policy.py`, `handles.py`,
`kernel.py`, `models.py`, the firewall) so doing them together avoided
three separate rounds of refactoring the same files. Per Mode B and
explicit owner permission, no backwards-compatibility shims were added
to the `Handle` dataclass.

## How verified

- `ruff format --check src/ tests/ examples/` ÔÇö 49 files already formatted.
- `ruff check src/ tests/ examples/` ÔÇö All checks passed.
- `python -m mypy src/` ÔÇö Success: no issues found in 30 source files.
- `python -m pytest -q --cov=agent_kernel` ÔÇö **477 passed** (was 450; +27
  new tests). Coverage 96%.
- `examples/basic_cli.py`, `examples/billing_demo.py`,
  `examples/http_driver_demo.py` ÔÇö all run clean.

## Tradeoffs / risks

- **No backwards compat on `Handle`** (per Mode B): the dataclass gained
  two new fields with safe defaults (empty `principal_id`, empty
  `constraints`). Direct `Handle(...)` constructors that relied on
  positional-only args would break, but every existing test passes
  unchanged because all callers use keyword args.
- **MEMORY rule placement**: explicitly placed *after* PII/PCI/SECRETS
  checks to honour the documented invariant ordering. New `SensitivityTag`
  enum values need a matching policy rule ÔÇö that rule was added in the
  same change.
- **Redaction is scoped to `memory.*` by capability-ID prefix**, not by
  sensitivity tag, because the input-side redaction needs to apply
  regardless of whether the caller set the sensitivity tag correctly.
  Non-memory caps are unchanged.
- **Out of scope**: did not decompose `policy_dsl.py` (issue #68 D ÔÇö file
  is now 661 lines, even bigger). Did not add an OTel exporter, marketplace,
  or memory backend. Each can land separately.

## Scope notes (Mode B)

Three issues, one PR. Adjacent issues left as separate follow-ups: #68
(tech debt + dry-run driver coverage), #38 (OTel), #45 (namespaces),
#46 (tutorial), #47 (streaming), #49/#51/#52 (marketplace), #71 (boundary
docs).

## Closes

Closes #74, closes #75, closes #76.

(#73 and #77 already closed manually during triage ÔÇö implemented by PR
#78 but auto-close didn't fire.)

### CODE REVIEW

File: docs/security.md [hunk 1]
```diff
@@ -27,6 +30,53 @@ Any change to these fields invalidates the HMAC signature.
 
 Consider an agent that obtains a token for `billing.list_invoices` then passes it to a different agent. The second agent cannot use it because `verify()` checks that `token.principal_id == expected_principal_id`.
 
+The same principle extends to handles: every `Handle` carries the `principal_id`
+the original grant was issued to, and `HandleStore.expand(..., principal_id=...)`
+rejects expansion attempts from any other principal with
+`HandleConstraintViolation` (`reason_code = HANDLE_PRINCIPAL_MISMATCH`). A
+handle ID alone is not a bearer credential.
```
[the remark this answers is not in the document]
Comment (dgenio) [reply docs/security.md]: Already addressed (GitHub also marked this outdated). The documentation at `docs/security.md` lines 36–40 explicitly states: "An omitted or empty `principal_id` is treated as a mismatch" — which matches the implementation.

File: src/agent_kernel/handles.py [hunk 2]
```diff
@@ -149,20 +161,65 @@ def expand(
...
+                ``scope``) or is requested by a different principal.
         """
+        # ── Principal binding ─────────────────────────────────────────────────
+        if handle.principal_id and principal_id and handle.principal_id != principal_id:
+            raise HandleConstraintViolation(
+                f"Handle '{handle.handle_id}' was granted to principal "
+                f"'{handle.principal_id}' and cannot be expanded by "
+                f"'{principal_id}'.",
+                reason_code=DenialReason.HANDLE_PRINCIPAL_MISMATCH,
+            )
```
[the remark this answers is not in the document]
Comment (dgenio) [reply src/agent_kernel/handles.py:188]: Already addressed. The condition `if handle.principal_id and handle.principal_id != principal_id:` correctly denies when `principal_id=""` (the default) because `"p-original" != ""` evaluates to `True`. The inline code comment explicitly states "omitting it counts as a mismatch." Test `test_expand_principal_missing_denied_for_bound_handle` covers both the omitted and explicit empty-string cases.

File: tests/test_handles.py [hunk 3]
```diff
@@ -145,3 +151,113 @@ def test_periodic_eviction_on_store() -> None:
...
+    handle = store.store(
+        "cap.x",
+        _granted_rows(),
+        principal_id="p-original",
+    )
+    with pytest.raises(HandleConstraintViolation) as exc:
+        store.expand(handle, query={}, principal_id="p-attacker")
+    assert exc.value.reason_code == DenialReason.HANDLE_PRINCIPAL_MISMATCH
+
+
```
[the remark this answers is not in the document]
Comment (dgenio) [reply tests/test_handles.py:252]: Already addressed. Test `test_expand_principal_missing_denied_for_bound_handle` (lines 256–272 of `tests/test_handles.py`) does exactly this — it asserts denial for both the omitted case (`store.expand(handle, query={})`) and the explicit empty string case (`principal_id=""`), with a regression note in the docstring explaining the earlier vulnerability.

File: src/agent_kernel/handles.py [hunk 4]
```diff
@@ -149,20 +161,65 @@ def expand(
...
         # ── Filtering ──────────────────────────────────────────────────────────
-        filter_spec: dict[str, Any] = query.get("filter", {})
-        if filter_spec and isinstance(filter_spec, dict):
+        # Grant scope is AND-merged into the request filter so the caller
+        # cannot bypass it by omitting the scope key.
+        filter_spec: dict[str, Any] = dict(query.get("filter", {}) or {})
+        for sk, sv in granted_scope.items():
+            filter_spec.setdefault(sk, sv)
+        if filter_spec:
             rows = [
```
[the remark this answers is not in the document]
Comment (dgenio) [reply src/agent_kernel/handles.py:245]: Already addressed. Lines 196–205 validate `filter` (must be `dict`) and `fields` (must be `list|tuple`) with explicit `isinstance` checks before they're used. Invalid types raise `HandleConstraintViolation(reason_code=INVALID_CONSTRAINT)` — not a bare stdlib exception.

File: src/agent_kernel/handles.py [hunk 5]
```diff
@@ -171,14 +228,35 @@ def expand(
 
         # ── Pagination ────────────────────────────────────────────────────────
         offset = int(query.get("offset", 0))
-        limit = int(query.get("limit", len(rows)))
+        requested_limit = query.get("limit")
+        limit = len(rows) if requested_limit is None else int(requested_limit)
+
+        if isinstance(granted_max_rows, int) and granted_max_rows >= 0:
+            if requested_limit is not None and int(requested_limit) > granted_max_rows:
+                raise HandleConstraintViolation(
```
[the remark this answers is not in the document]
Comment (dgenio) [reply src/agent_kernel/handles.py:275]: Already addressed. Both `offset` and `limit` conversions are wrapped in `try/except (TypeError, ValueError) as exc` and remapped to `HandleConstraintViolation(reason_code=INVALID_CONSTRAINT)` with a stable error message. No bare stdlib exceptions propagate.
